### ISSUE TITLE
Use default Speaker when voiceover is disabled

### ISSUE THREAD

**Issue by NicoHood** [nims11/IPod-Shuffle-4g#17]
Normally you hear a male speaker if no voiceover was selected via ITunes. Those sound files are still there `./NICO'S IPOD/iPod_Control/Speakable/Messages/`.

However I checked my old Ipod backup and it seems the ipod does not generate any database for this "non voiceover" ipod settings from itunes. So there must be a way to use the internal voices without merging the sounds and doing complicated stuff.

If I disable voiceover with your tool I only get "Playlist zero" for ALL playlists. "All Songs" disappeared and is silent.

We should be able to fix this somehow. This also means we do not have to provide the closed source male speaker voiceover, which is great. I will look at the hex output with `xxd`. But I have nothing more than my backup. Not even a windows PC I could use to regenerate Itunes files.

**Comment by NicoHood** [comment]
Partly solved here:
https://github.com/NicoHood/IPod-Shuffle-4g/commit/78482c9aebcd51d6035b602091b7c5be6f1105a0

The Playlist n is just missing. I tried to set the hex value to the following, but it always outputs "zero"
```
"\0\0\0\0\0\0\0\1"
"\0\0\0\0\0\0\0\2"
"\0\0\0\2\0\0\0\0"
"\2\0\0\0\0\0\0\0"
```

Information was taken from this section:
https://github.com/nims11/IPod-Shuffle-4g/blob/master/docs/iTunesSD3gen.md#playlistx

If I can get the playlist n to work I will open another PR.

**Comment by NicoHood** [comment]
Maybe we can extract more information from here:
https://github.com/GNOME/rhythmbox/tree/master/plugins/ipod

I still cannot find the playlist fake voiceover yet. But we might get additional track field information from this, like stars (rating) and others. However it should be useless anyways for an ipod shuffle.

**Comment by NicoHood** [comment]
The playlist header format is totally different to the docs:
https://github.com/nims11/IPod-Shuffle-4g/blob/master/docs/iTunesSD3gen.md#playlist-header

The same playlists (form above) now generated with the script gives:
```
00001c0: 0000 0000 0000 0000 0000 0000 6870 6873  ............hphs
00001d0:|4c00 0000|0200 0000|ffff|ffff|0000|0000| L...............
00001e0: ffff ffff 0000 0000 ffff ffff 0000 0000  ................
00001f0: ffff ffff 0000 0000 ffff ffff 0000 0000  ................
0000200: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000210: 1802 0000 4802 0000 6c70 6873 3000 0000  ....H...lphs0...
0000220: 0100 0000 0100 0000 0000 0000 0000 0000  ................
0000230: 0100 0000 0000 0000 0000 0000 0000 0000  ................
0000240: 0000 0000 0000 0000 6c70 6873 3000 0000  ........lphs0...
0000250: 0100 0000 0100 0000 1db1 7926 8a06 9bec  ..........y&....
0000260: 0200 0000 0000 0000 0000 0000 0000 0000  ................
0000270: 0000 0000 0000 0000                      ........
```

You can see there is very weird data in it. I think the bytelength is wrong and also the information inside. And thatswhy the ipod outputs zero probably. Not because we have a playlist header problem, I think its more a playlist header header problem.

**Comment by NicoHood** [comment]
I found a Rythmbox bug X_x

If you sync 2 playlists with 1 and x songs, it will transmit only 1, x-1 songs as playlist. This first confused me.

I got the header now setup exactly as rythmbox, except the voiceover filename (which i dont know if its random). However my ipod crashes if i hit the playlist button. It keeps playing music, even after turning off. After a while it then goes off luckly. Nice to see it crashing like that. If I dont destroy the aux plugs, then I will do it by software ;D

This confuses me even more. Why does a fixed header not work and this unknown one does!? There must be something wrong still.

See my headers (last playlist misses 1 song as described)
```hex
00004b0: 0000 0000 0000 0000 0000 0000 6870 6873  ............hphs
00004c0: 2000 0000 0300 0000 0300 0100 0300 0000   ...............
00004d0: dc04 0000 1405 0000 4405 0000 6c70 6873  ........D...lphs
00004e0: 3800 0000 0300 0000 0300 0000 0000 0000  8...............
00004f0: 0000 0000 0100 0000 0000 0000 0000 0000  ................
0000500: 0000 0000 0000 0000 0000 0000 0100 0000  ................
0000510: 0200 0000 6c70 6873 3000 0000 0100 0000  ....lphs0.......
0000520: 0100 0000 1db1 7926 8a06 9bec 0200 0000  ......y&........
0000530: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000540: 0200 0000 6c70 6873 3000 0000 0100 0000  ....lphs0.......
0000550: 0100 0000 3b43 3a8c 1332 b6ad 0200 0000  ....;C:..2......
0000560: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000570: 0100 0000                                ....
```

**Comment by NicoHood** [comment]
As reference what i did so far:

```diff
IPod-Shuffle-4g$ git diff
diff --git a/shuffle.py b/shuffle.py
index 78c9719..82617d1 100755
--- a/shuffle.py
+++ b/shuffle.py
@@ -281,7 +281,7 @@ class Track(Record):
                            ("track", ("H", 1)),
                            ("disc", ("H", 0)),
                            ("unknown4", ("Q", 0)),
-                           ("dbid", ("8s", 0)),
+                           ("dbid", ("8s", "\x00" * 8)),
                            ("artistid", ("I", 0)),
                            ("unknown5", ("32s", "\x00" * 32)),
                            ])
@@ -328,16 +328,10 @@ class PlaylistHeader(Record):
                           ("header_id", ("4s", "shph")),
                           ("total_length", ("I", 0)),
                           ("number_of_playlists", ("I", 0)),
-                          ("number_of_podcast_lists", ("I", 0xffffffff)),
-                          ("number_of_master_lists", ("I", 0)),
-                          ("number_of_audiobook_lists", ("I", 0xffffffff)),
-                          ("unknown1", ("I", 0)),
-                          ("unknown2", ("I", 0xffffffff)),
-                          ("unknown3", ("I", 0)),
-                          ("unknown4", ("I", 0xffffffff)),
-                          ("unknown5", ("I", 0)),
-                          ("unknown6", ("I", 0xffffffff)),
-                          ("unknown7", ("20s", "\x00" * 20)),
+                          ("number_of_non_podcast_lists", ("2s", "\x03\x00")), #TODO check if really ffff is okay
+                          ("number_of_master_lists", ("2s", "\x01\x00")),
+                          ("number_of_non_audiobook_lists", ("2s", "\x03\x00")), #TODO as above
+                          ("unknown2", ("2s", "\x00" * 2)),
                                               ])
 
     def construct(self, tracks): #pylint: disable-msg=W0221
@@ -359,8 +353,7 @@ class PlaylistHeader(Record):
                 chunks += [construction]
 
         self["number_of_playlists"] = playlistcount
-        self["number_of_master_lists"] = 0
-        self["total_length"] = 0x44 + (self["number_of_playlists"] * 4)
+        self["total_length"] = 0x14 + (self["number_of_playlists"] * 4)
         # Start the header
 
         output = Record.construct(self)
```

Any thoughs appreciated. But I will continue self conversion here, sometimes this helps. And others can pick up the work if I ever give up. But not yet time for it ;)

**Comment by nims11** [comment]
Ah, quite some work done by you! Will take me some time to look into it and relate to your findings (Weekdays -_-). Give me a day or two to get back to you on this.

**Comment by NicoHood** [comment]
I found the bug!!!!

I applied the patch I showed above. I do not know where this weird data is coming from, but it does not apply the ipod shuffle specs we have in our documents. Maybe this is an older ipod shuffle gen 3 format? I do not know. And I do not own such a device, and I also think we should not care. The rythmbox plugin handles it the same way.

The critical difference however is the 2nd 32bit value of the database.
https://github.com/nims11/IPod-Shuffle-4g/blob/e02ef0fff658990238625c10bb0150be4d60e3a0/shuffle.py#L192
Also our docs list a different value here (the one I fixed in the PR). I do not know where this also comes from, but its wrong. And this makes the ipod not work with the built-in sound files. I do not know why this happens, but can simply fix it.

I still do not have a windows machine to test an itunes setting. Just a little note here:
Windows 10 in vm does not recognize the USB device.
I then tried Windows 7. But it does stop finding updates after SP1. It just searches forever with 100% cpu. I thought this was caused by the vm, but it also happened to someone else in vm and real hardware. See what windows is doing here to break fresh win7 installations to force us use win10?
So no way for me to test itunes.

PR will come soon.

**Comment by NicoHood** [comment]
Also see this:
http://sourceforge.net/p/gtkpod/mailman/message/32333835/

It has to be a version number.

**Comment by nims11** [comment]
I tried syncing 3 files from my itunes in windows. Turns out it generates a completely different database format (mhbd, different from the bdhs we are using). I was able to find some references on this format (http://homepage.ntlworld.com/simon.mason20/ipod_tunes_spec.htm), but it seems incomplete. Will probe and update on it.
[iTunesDB.zip](https://github.com/nims11/IPod-Shuffle-4g/files/102404/iTunesDB.zip)

**Comment by NicoHood** [comment]
We are looking at the iTunesSD file not iTunesDB file. That might be the difference. Please upload the whole iTunes folder.

**Comment by nims11** [comment]
Ah, hit me, sorry for that :|
[iTunesSD.zip](https://github.com/nims11/IPod-Shuffle-4g/files/102486/iTunesSD.zip)

**Comment by NicoHood** [comment]
Differences:
* `unknown_3` ** is 1 instead of 0.**
* `stop_at_pos_ms` seem to differ for 2/3 songs?? (rythmbox doesnt use this at all)
* dbid is different, but expected.
* `artist id`, `track number`, `album id` is different, might be some itunes internals, nothing important. But we can read this from id3 tag if we want to. But not all files have this. Probably set the numbers all to zero then.
* **we use a post and pregap of 2 which i was wondering about.**

Is the post/pregap important? Is it something like crossfade or a cutoff? We might want to test that. I can manipulate my database.

I am just wondering about unknown_3 value...

You used my PR to generate the 2nd data? Otherwise it must have been different.

FYI:
I use `xxd` to dump the data, `meld` to compare and `bless` to manipulate it.

**Comment by NicoHood** [comment]
Also see this:
http://sourceforge.net/p/gtkpod/code/ci/master/tree/itunesdb_info.txt

**Comment by nims11** [comment]
Fixed f7eafd1
